File: packages/coverage/groma_coverage/kernel.py
"""The coverage kernel. Build spec 6.3, 6.4.

Pure. NumPy in, NumPy out. No I/O, no database, no framework, no logging of user
data, so that it runs identically in the worker, the CLI and the browser via WASM.

KERNEL_VERSION is recorded on every coverage_run and printed in every report,
because these numbers end up in tender documents. Bump it for ANY behavioural
change, and explain the movement in the commit message before touching a golden
file.
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import Final

import numpy as np

from groma_contracts.camera import CameraSpec
from groma_coverage.occluders import (
    RAY_EPS_M,
    PreparedOccluder,
    SegmentBatch,
    prepare,
)
from groma_coverage.terrain import eval_heights, terrain_blocks
from groma_coverage.types import CoverageResult, Grid, Occluder, Terrain
from groma_geo.optics import camera_basis, f_px, hfov_rad, vfov_rad

KERNEL_VERSION: Final[str] = "1.1.0"


def compute_coverage(
    cameras: Sequence[CameraSpec],
    occluders: Sequence[Occluder],
    grid: Grid,
    terrain: Terrain | None = None,
    eval_height_m: float = 1.6,
    foreshorten: bool = True,
) -> CoverageResult:
    """Best achievable pixel density per grid cell, across all enabled cameras.

    `eval_height_m` is above the local terrain surface, not above the datum. Where
    `terrain` is None the ground is the plane y = 0. The height actually used is
    recorded per cell in `CoverageResult.eval_y`, so a report can state it.

    Cameras are indexed by their position in `cameras`, including disabled ones, so
    that `best_camera` indices stay stable when a camera is toggled in the UI.
    """
    nz, nx = grid.nz, grid.nx

    xs, zs = grid.centres()
    ys = eval_heights(xs, zs, terrain, eval_height_m)

    # (cells, 3) targets. Flattened once; every per-camera pass reuses this.
    targets = np.stack([xs.ravel(), ys.ravel(), zs.ravel()], axis=1)
    n_cells = targets.shape[0]

    ppm = np.zeros(n_cells, dtype=np.float64)
    count = np.zeros(n_cells, dtype=np.uint8)
    best = np.full(n_cells, -1, dtype=np.int16)

    prepared = [prepare(o) for o in occluders]

    for index, cam in enumerate(cameras):
        if not cam.enabled:
            continue
        if cam.roll_deg != 0.0:
            raise NotImplementedError(
                f"camera {cam.id!r} has roll_deg={cam.roll_deg}; the frustum test is "
                "roll-free (groma_geo.optics.camera_basis). Rolled frusta need the "
                "basis rotated about forward and the analytic tests extended."
            )

        cam_ppm, cam_seen = _coverage_one_camera(cam, prepared, targets, terrain, foreshorten)

        improved = cam_seen & (cam_ppm > ppm)
        ppm = np.where(improved, cam_ppm, ppm)
        best = np.where(improved, np.int16(index), best)
        count += cam_seen.astype(np.uint8)

    return CoverageResult(
        ppm=ppm.reshape(nz, nx).astype(np.float32),
        count=count.reshape(nz, nx),
        best_camera=best.reshape(nz, nx),
        eval_y=ys.reshape(nz, nx).astype(np.float32),
        grid=grid,
        kernel_version=KERNEL_VERSION,
    )


def _coverage_one_camera(
    cam: CameraSpec,
    prepared: Sequence[PreparedOccluder],
    targets: np.ndarray,
    terrain: Terrain | None,
    foreshorten: bool,
) -> tuple[np.ndarray, np.ndarray]:
    """Return (ppm, seen) over all cells for one camera.

    `seen` is the camera-count contribution: a cell is seen if it is in range, in
    the frustum, and not fully blocked. A cell behind a porous occluder is seen,
    at a reduced pixel density.
    """
    n_cells = targets.shape[0]
    ppm = np.zeros(n_cells, dtype=np.float64)
    seen = np.zeros(n_cells, dtype=bool)

    origin = np.array([cam.position.x, cam.position.y, cam.position.z], dtype=np.float64)

    # Step 1: V = P - camera.position, and squared range.
    v = targets - origin
    d2 = np.einsum("ij,ij->i", v, v)

    # Step 2: range gate.
    mask = (d2 >= cam.near_m * cam.near_m) & (d2 <= cam.far_m * cam.far_m)
    if not np.any(mask):
        return ppm, seen

    # Step 3: frustum, in the camera basis.
    forward, right, up = camera_basis(cam.pan_deg, cam.tilt_deg)
    tan_h = np.tan(hfov_rad(cam.sensor_w_mm, cam.focal_mm) / 2.0)
    tan_v = np.tan(vfov_rad(cam.sensor_h_mm, cam.focal_mm) / 2.0)

    zc = v @ forward
    mask &= zc > 0.0
    if not np.any(mask):
        return ppm, seen
    xc = v @ right
    yc = v @ up
    mask &= (np.abs(xc) <= zc * tan_h) & (np.abs(yc) <= zc * tan_v)

    # Step 4: reduce to surviving candidates. Typically 10-40% of cells, and every
    # occlusion test below is proportional to this count.
    cand = np.flatnonzero(mask)
    if cand.size == 0:
        return ppm, seen

    cand_targets = targets[cand]
    d = np.sqrt(d2[cand])
    eps_t = np.minimum(RAY_EPS_M / np.maximum(d, RAY_EPS_M), 0.49)

    # Step 5: occlusion. Solid occluders block; porous ones attenuate.
    transmission = np.ones(cand.size, dtype=np.float64)
    blocked = np.zeros(cand.size, dtype=bool)
    batch = SegmentBatch.build(origin, cand_targets)

    for occ in prepared:
        # Exclusion: a camera is never occluded by the structure it is mounted on.
        # Half of the self-occlusion fix; the other half is the bracket offset,
        # already applied to cam.position (explained 7.7, T8).
        if cam.mount_structure_id is not None and occ.owner_id == cam.mount_structure_id:
            continue
        if occ.transmission >= 1.0:
            continue

        near = occ.broad_phase(batch)
        live = near & ~blocked
        if not np.any(live):
            continue

        idx = np.flatnonzero(live)
        hit = occ.hits(origin, cand_targets[idx], eps_t[idx])
        if not np.any(hit):
            continue

        struck = idx[hit]
        if occ.solid:
            blocked[struck] = True
        else:
            transmission[struck] *= occ.transmission

    # Step 6: terrain occlusion.
    if terrain is not None:
        live = ~blocked
        if np.any(live):
            idx = np.flatnonzero(live)
            ground_hit = terrain_blocks(origin, cand_targets[idx], terrain, eps_t[idx])
            blocked[idx[ground_hit]] = True

    visible = ~blocked
    if not np.any(visible):
        return ppm, seen

    # Step 7: pixel density.
    f = f_px(cam.focal_mm, cam.res_y, cam.sensor_h_mm)
    value = f / d
    if foreshorten:
        # cos of the depression angle: the target is a vertical surface, so a ray
        # arriving steeply from above sees a foreshortened face.
        sin_dep = np.abs(v[cand, 1]) / d
        value = value * np.sqrt(np.maximum(0.0, 1.0 - sin_dep * sin_dep))
    value = value * transmission

    out_idx = cand[visible]
    ppm[out_idx] = value[visible]
    seen[out_idx] = True
    return ppm, seen


__all__ = ["KERNEL_VERSION", "compute_coverage"]

File: apps/web/src/kernel/kernel.ts
/* groma_coverage.kernel. Bump KERNEL_VERSION together with the Python one. */

import { hits, prepare, RAY_EPS_M, type Prepared } from "./occluders";
import { cameraBasis, fPx, hfovRad, vfovRad } from "./optics";
import { terrainBlocks } from "./terrain";
import { gridNx, gridNz, terrainHeightAt, terrainMax, type Camera, type CoverageResult, type Grid, type Occluder, type Terrain } from "./types";

export const KERNEL_VERSION = "1.1.0";

export function computeCoverage(
  cameras: Camera[],
  occluders: Occluder[],
  grid: Grid,
  terrain: Terrain | null,
  evalHeightM = 1.6,
  foreshorten = true,
): CoverageResult {
  const nx = gridNx(grid), nz = gridNz(grid), n = nx * nz;
  const xs = new Float64Array(n), ys = new Float64Array(n), zs = new Float64Array(n);
  for (let j = 0; j < nz; j++)
    for (let i = 0; i < nx; i++) {
      const k = j * nx + i;
      const x = grid.x_min + i * grid.spacing, z = grid.z_min + j * grid.spacing;
      xs[k] = x; zs[k] = z;
      ys[k] = (terrain ? terrainHeightAt(terrain, x, z) : 0) + evalHeightM;
    }
  const ppm = new Float64Array(n);
  const count = new Uint8Array(n);
  const best = new Int16Array(n).fill(-1);
  const prepared: Prepared[] = occluders.map(prepare);
  const yMax = terrain ? terrainMax(terrain) : -Infinity;

  cameras.forEach((cam, index) => {
    if (cam.enabled === false) return;
    if ((cam.roll_deg ?? 0) !== 0) throw new Error(`camera ${cam.id} has roll; the frustum test is roll-free`);
    const ox = cam.position.x, oy = cam.position.y, oz = cam.position.z;
    const near = cam.near_m ?? 1, far = cam.far_m ?? 200;
    const near2 = near * near, far2 = far * far;
    const { forward: f, right: r, up: u } = cameraBasis(cam.pan_deg, cam.tilt_deg);
    const tanH = Math.tan(hfovRad(cam.sensor_w_mm, cam.focal_mm) / 2);
    const tanV = Math.tan(vfovRad(cam.sensor_h_mm, cam.focal_mm) / 2);
    const fpx = fPx(cam.focal_mm, cam.res_y, cam.sensor_h_mm);
    const mount = cam.mount_structure_id ?? null;
    const occ = prepared.filter((p) => !p.transparent && !(mount !== null && p.ownerId === mount));

    for (let k = 0; k < n; k++) {
      const vx = xs[k]! - ox, vy = ys[k]! - oy, vz = zs[k]! - oz;
      const d2 = vx * vx + vy * vy + vz * vz;
      if (d2 < near2 || d2 > far2) continue;
      const zc = vx * f[0] + vy * f[1] + vz * f[2];
      if (zc <= 0) continue;
      const xc = vx * r[0] + vy * r[1] + vz * r[2];
      if (Math.abs(xc) > zc * tanH) continue;
      const yc = vx * u[0] + vy * u[1] + vz * u[2];
      if (Math.abs(yc) > zc * tanV) continue;
      const d = Math.sqrt(d2);
      const epsT = Math.min(RAY_EPS_M / Math.max(d, RAY_EPS_M), 0.49);
      let transmission = 1;
      let blocked = false;
      for (const p of occ) {
        if (!hits(p, ox, oy, oz, xs[k]!, ys[k]!, zs[k]!, epsT)) continue;
        if (p.solid) { blocked = true; break; }
        transmission *= p.porosity;
      }
      if (blocked) continue;
      if (terrain && terrainBlocks(terrain, yMax, ox, oy, oz, xs[k]!, ys[k]!, zs[k]!, epsT)) continue;
      let value = fpx / d;
      if (foreshorten) {
        const sinDep = Math.abs(vy) / d;
        value *= Math.sqrt(Math.max(0, 1 - sinDep * sinDep));
      }
      value *= transmission;
      count[k] = count[k]! + 1;
      if (value > ppm[k]!) { ppm[k] = value; best[k] = index; }
    }
  });

  return { ppm: Float32Array.from(ppm), count, best_camera: best, eval_y: Float32Array.from(ys), nx, nz, grid, kernel_version: KERNEL_VERSION };
}
